src/sortMachOFiles.js

```diff
@@ -0,0 +1,241 @@
+import { open, read } from 'fs';
+
+const MH_MAGIC_64 = 0xfeedfacf;
+const MH_CIGAM_64 = 0xcffaedfe;
+const MH_CIGAM = 0xcefaedfe;
+const FAT_MAGIC = 0xcafebabe;
+const FAT_CIGAM = 0xbebafeca;
+
+const CPU_TYPE_I386 = 0x00000007;
+const CPU_TYPE_X86_64 = 0x01000007;
+const CPU_TYPE_ARM = 0x0000000c;
+const CPU_TYPE_ARM64 = 0x0100000c;
+
+const ARCH_NAMES = {
+  [CPU_TYPE_ARM]: 'arm',
+  [CPU_TYPE_ARM64]: 'arm64',
+  [CPU_TYPE_I386]: 'i386',
+  [CPU_TYPE_X86_64]: 'x86_64',
+};
+
+const LC_UUID = 0x1b;
+const UUID_BYTES = 16;
+
+const MAGIC_NUMBER_BYTES = 4;
+const MACH_HEADER_BYTES = 28;
+const MACH_64_HEADER_BYTES = 32;
+
+const FAT_ARCH_BYTES = 20;
+const FAT_HEADER_BYTES = 8;
+
+const LOAD_COMMAND_BYTES = 8;
+
+function readBytes(fd, position, length) {
+  return new Promise((resolve, reject) => {
+    const output = Buffer.alloc(length);
+    read(fd, { buffer: output, length, position }, (err) => err ? reject(err) : resolve(output));
+  });
+}
+
+function getMagic(buffer) {
+  const view = new DataView(buffer.buffer);
+  return view.getUint32(0);
+}
+
+function getIsMach64Header(magic) {
+  return magic === MH_MAGIC_64 || magic === MH_CIGAM_64;
+}
+
+function shouldSwapBytes(magic) {
+  return magic === MH_CIGAM || magic === MH_CIGAM_64;
+}
+
+function getIsMultiArch(magic) {
+  return magic === FAT_CIGAM || magic === FAT_MAGIC;
+}
+
+function getNumArchs(buffer) {
+  const bufDataView = new DataView(buffer.buffer);
+  return bufDataView.getUint32(4);
+}
+
+function getArchDetails(buffer) {
+  const bufDataView = new DataView(buffer.buffer);
+  return {
+    cpuType: bufDataView.getUint32(0),
+    archOffset: bufDataView.getUint32(8),
+  };
+}
+
+function getHeaderVals(buffer, shouldSwap) {
+  const buf = shouldSwap ? buffer.reverse().buffer : buffer.buffer;
+  const bufDataView = new DataView(buf);
+  return {
+    cpuType: bufDataView.getUint32(4),
+    ncmds: bufDataView.getUint32(16),
+  };
+}
+
+function getLoadCommand(buffer, shouldSwap) {
+  const buf = shouldSwap ? buffer.reverse().buffer : buffer.buffer;
+  const bufDataView = new DataView(buf);
+
+  let cmdSize;
+  let cmd;
+
+  if (shouldSwap) {
+    cmdSize = bufDataView.getUint32(0);
+    cmd = bufDataView.getUint32(4);
+  } else {
+    cmdSize = bufDataView.getUint32(4);
+    cmd = bufDataView.getUint32(0);
+  }
+  return { cmd, cmdSize };
+}
+
+async function getArchEntry(fd, archOffset, magic, archName = null) {
+  const isMach64Header = getIsMach64Header(magic);
+  const headerSize = isMach64Header ? MACH_64_HEADER_BYTES : MACH_HEADER_BYTES;
+  const shouldSwap = shouldSwapBytes(magic);
+
+  let headerBuffer;
+  try {
+    headerBuffer = await readBytes(fd, archOffset, headerSize);
+  } catch (err) {
+    throw ['Error reading header', err]
```

Comment:
While you _can_ throw an array, this is not a great way to handle exceptions in javascript. First off, the error handling is `console.error(anotherMessage, thisError)`, which will print something like:

```
Another message [The wrapped message, Error: wrapped exception
    at foo:1:2]
```

and yet other calls to the error handler receive an actual error object, and so will print:

```
Another message Error: wrapped exception
  at foo:1:2
```

This is inconsistent and not the easiest to parse at a glance. For wrapping an "original error" I find this pattern to work well:

```
class WrappedError extends Error {
  constructor(msg, originalError) {
    super(msg);
    this.originalError = originalError;
  }

  get verboseMessage() {
    return `${this.message}\n\nOriginal error: ${this.originalError.stack}`;
  }
}
```

Now we have an actual Error type (matching what internal node systems would be returning, e.g. from `read`) and have a bit nicer of a format for our "simple message" as well as the original error contained.

By default (in non-verbose mode) we would just log the `error.message` (e.g. just "Error reading header" here), but in verbose mode we would instead log `verboseMessage` to provide further context if needed.

Our "simple message" here should also include _what file_ we are failing on - this is important context if something is a problem.